app/routes/checkout/billing-details.tsx

```diff
@@ -0,0 +1,92 @@
+import { Link, Form, useActionData } from '@remix-run/react'
+import { json, redirect } from '@remix-run/node'
+import { Text, TextInput, Button } from '~/components'
+import useMaskedInput from '~/hooks/use-masked-input'
+import { signUp } from '~/models/auth'
+import * as SessionStorage from '~/utils/session-storage'
+import formDataToObject from '~/utils/form-data-to-object'
+import RequestError from '~/errors/request-error'
+import { PHONE_REGEXP } from '~/constants'
+
+import type { ActionArgs } from '@remix-run/node'
+import type { UserInfoPayload } from '~/models/auth'
+
+interface ValidationErrors {
+  general?: string
+  name?: string
+  email?: string
+  phone?: string
+}
+
+const validateForm = (userInfo: UserInfoPayload) => {
+  const errors: ValidationErrors = {}
+  if (!userInfo.name) errors.name = 'Please enter your name'
+  if (!userInfo.email) errors.email = 'Please enter your email'
+  if (!userInfo.phone) errors.phone = 'Please enter your phone'
+  if (!PHONE_REGEXP.test(userInfo.phone)) errors.phone = 'Please enter a valid nubmer'
+
+  return errors
+}
+
+export const action = async ({ request }: ActionArgs) => {
+  const formData = await request.formData()
+  const userInfo = formDataToObject(formData) as UserInfoPayload
+  const { sessionId } = await SessionStorage.getOrCreateSessionId(request)
+
+  const errors = validateForm(userInfo)
+  if (Object.keys(errors).length) return json({ errors })
+
+  try {
+    await signUp(sessionId, userInfo)
+  } catch (error) {
+    if (!(error instanceof RequestError)) throw error
+    errors.general = error.message
+  }
+
+  if (Object.keys(errors).length) return json({ errors })
+
+  return redirect(`/checkout/confirmation-code?email=${encodeURIComponent(userInfo.email)}`)
+}
+
+export default () => {
+  const result = useActionData<typeof action>()
+  const errors: ValidationErrors = result ? result.errors : {}
```

Comment:
this is the billing details (a.k.a sign up) screen. The `<Form>` component below is provided by remix, remix itself will take care of submitting the form data, and it will execute what the `action` function defined on this same file. So, the `action` function in these route files is something similar to the controller in an MVC app.